File: packages/react-native-sdk/android/src/main/java/com/streamio/aicomponents/renderkit/AIDictationModule.kt
package com.streamio.aicomponents.renderkit

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.speech.RecognitionListener
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer
import androidx.core.content.ContextCompat
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.util.Locale
import java.util.concurrent.atomic.AtomicBoolean

class AIDictationModule(
    private val reactContext: ReactApplicationContext,
) : NativeAIDictationSpec(reactContext), LifecycleEventListener {

    // ---------------------------------------------------------------------------
    // Module constants
    // ---------------------------------------------------------------------------

    companion object {
        const val NAME = "NativeAIDictation"

        private const val EVENT_RESULT = "AIDictationResult"
        private const val EVENT_STATE = "AIDictationState"
        private const val EVENT_ERROR = "AIDictationError"

        private const val STATE_IDLE = "idle"
        private const val STATE_STARTING = "starting"
        private const val STATE_LISTENING = "listening"
        private const val STATE_STOPPING = "stopping"

        private const val ERROR_NO_PERMISSION = "NO_PERMISSION"
        private const val ERROR_NOT_AVAILABLE = "NOT_AVAILABLE"
        private const val ERROR_INTERNAL = "INTERNAL_ERROR"
        private const val ERROR_ANDROID = "ANDROID_ERROR"

        private const val DEFAULT_SILENCE_TIMEOUT_MS = 2500L
    }

    private val mainHandler = Handler(Looper.getMainLooper())
    private var speechRecognizer: SpeechRecognizer? = null

    private val isListening = AtomicBoolean(false)
    private var interimResultsEnabled: Boolean = true
    private var currentLocale: Locale = Locale.getDefault()
    private var silenceTimeoutMs: Long = DEFAULT_SILENCE_TIMEOUT_MS
    private var lastSpeechTimeMs: Long = 0L

    private var silenceRunnable: Runnable? = null

    override fun getName(): String = NAME

    override fun initialize() {
        super.initialize()
        reactContext.addLifecycleEventListener(this)
    }

    override fun invalidate() {
        super.invalidate()
        reactContext.removeLifecycleEventListener(this)
        cleanupRecognizer()
    }

    // ---------------------------------------------------------------------------
    // NativeAIDictationSpec API implementation
    // ---------------------------------------------------------------------------

    /**
     * Starts a new dictation session with the given options.
     *
     * - No-ops if a session is already active
     * - Verifies the RECORD_AUDIO permission and speech recognition availability before starting
     * - On success, configures and starts [SpeechRecognizer] as well as the silence monitor
     *
     * @param options Optional configuration map:
     * - `interimResults` (boolean, default `true`) – whether to emit partial/interim results
     * - `language` (string, BCP-47 tag) – language/locale for recognition; falls back to
     *   [Locale.getDefault] if omitted or blank
     * - `silenceTimeoutMs` (number) – max allowed silence in milliseconds before
     *   auto-stopping; falls back to [DEFAULT_SILENCE_TIMEOUT_MS] if missing or invalid
     */
    @ReactMethod
    override fun start(options: ReadableMap?) {
        if (isListening.get()) return

        emitState(STATE_STARTING)

        interimResultsEnabled = options?.getBooleanSafe("interimResults") ?: true

        val languageTag = options?.getStringSafe("language")
        currentLocale = if (languageTag.isNullOrBlank()) {
            Locale.getDefault()
        } else {
            Locale.forLanguageTag(languageTag)
        }

        val silenceOptMs = options?.getDoubleSafe("silenceTimeoutMs")
        silenceTimeoutMs = if (silenceOptMs != null && silenceOptMs > 0) {
            silenceOptMs.toLong()
        } else {
            DEFAULT_SILENCE_TIMEOUT_MS
        }

        if (!hasAudioPermission()) {
            emitError(
                ERROR_NO_PERMISSION,
                "RECORD_AUDIO permission not granted. Request it in JS before calling start().",
            )
            emitState(STATE_IDLE)
            return
        }

        if (!SpeechRecognizer.isRecognitionAvailable(reactContext)) {
            emitError(
                ERROR_NOT_AVAILABLE,
                "Speech recognition is not available on this device.",
            )
            emitState(STATE_IDLE)
            return
        }

        mainHandler.post {
            try {
                setupRecognizerIfNeeded()

                val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
                    putExtra(
                        RecognizerIntent.EXTRA_LANGUAGE_MODEL,
                        RecognizerIntent.LANGUAGE_MODEL_FREE_FORM,
                    )
                    putExtra(RecognizerIntent.EXTRA_LANGUAGE, currentLocale.toLanguageTag())
                    putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, interimResultsEnabled)
                    putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 3)
                    putExtra(RecognizerIntent.EXTRA_PROMPT, "Listening...")
                }

                lastSpeechTimeMs = System.currentTimeMillis()
                isListening.set(true)
                emitState(STATE_LISTENING)

                speechRecognizer?.startListening(intent)
                startSilenceMonitor()
            } catch (t: Throwable) {
                isListening.set(false)
                emitState(STATE_IDLE)
                emitError(
                    ERROR_INTERNAL,
                    "Failed to start recognition: ${t.message ?: t.javaClass.simpleName}",
                )
            }
        }
    }

    /**
     * Requests a graceful stop of the current dictation session.
     *
     * - No-ops if we are not currently listening
     * - Emits [STATE_STOPPING] so JS can show a "stopping" state
     * - Stops the silence monitor
     * - On the main thread, calls [SpeechRecognizer.stopListening] to let
     *   Android finish the current utterance and deliver final results
     *
     * The recognizer will transition to [STATE_IDLE] via [onResults] or [onError].
     */
    @ReactMethod
    override fun stop() {
        if (!isListening.get()) return

        emitState(STATE_STOPPING)
        stopSilenceMonitor()

        mainHandler.post {
            try {
                speechRecognizer?.stopListening()
            } catch (_: Throwable) {
                // ignore
            }
        }
    }

    /**
     * Immediately cancels the current dictation session.
     *
     * - No-ops if we are not currently listening
     * - Stops the silence monitor
     * - On the main thread, calls [SpeechRecognizer.cancel] to abort recognition
     * - Regardless of the cancel outcome, marks `isListening` false and emits
     *   [STATE_IDLE] so JS knows dictation has fully stopped
     */
    @ReactMethod
    override fun cancel() {
        if (!isListening.get()) return

        stopSilenceMonitor()

        mainHandler.post {
            try {
                speechRecognizer?.cancel()
            } catch (_: Throwable) {
                // ignore
            } finally {
                isListening.set(false)
                emitState(STATE_IDLE)
            }
        }
    }

    /**
     * Synchronously reports whether a dictation session is currently active.
     *
     * Exposed as a blocking synchronous method to JS so callers can quickly check
     * recording state without going through an async event.
     */
    @ReactMethod(isBlockingSynchronousMethod = true)
    override fun isRecording(): Boolean {
        return isListening.get()
    }

    // ---------------------------------------------------------------------------
    // React Native host lifecycle callbacks
    // ---------------------------------------------------------------------------


    /**
     * Called when the React Native host Activity is resumed (app comes to foreground).
     *
     * We currently don’t need to resume or re-attach anything for dictation here,
     * so this is intentionally a no-op.
     */
    override fun onHostResume() = Unit

    /**
     * Called when the React Native host Activity is paused (app goes to background
     * or another Activity is shown on top).
     *
     * If we are actively listening for speech when this happens, we cancel the
     * current dictation session. This avoids:
     *
     * - Keeping the microphone active while the app is in the background
     * - Letting the platform speech recognizer run in an invalid/paused state
     *
     * `cancel()` is used instead of `stop()` so the partial utterance is treated as
     * aborted and not as a successfully completed transcription.
     */
    override fun onHostPause() {
        if (isListening.get()) {
            cancel()
        }
    }

    /**
     * Called when the React Native host Activity is being destroyed.
     *
     * This is the last chance to clean up any native resources held by the
     * speech recognizer (listeners, microphone, underlying platform objects).
     * Technically, we typically want to do this within React Native code itself,
     * (i.e hook cleanup) however that might not always be enough in certain
     * scenarios and we keep this as a safety measure to make sure any memory leaks
     * don't occur.
     */
    override fun onHostDestroy() {
        cleanupRecognizer()
    }

    // ---------------------------------------------------------------------------
    // Internal
    // ---------------------------------------------------------------------------

    /**
     * Creates (if needed) and configures the Android [SpeechRecognizer] instance.
     *
     * The recognizer is bound to the current [reactContext] and wired to forward
     * Android recognition callbacks into the JavaScript side via:
     *
     * - [emitResult] for partial and final transcription text
     * - [emitAndroidError] for platform error codes
     * - [emitState] to update the high-level dictation state (i.e IDLE vs LISTENING)
     *
     * Callback behavior:
     *
     * - [RecognitionListener.onError]:
     *   - Cleans up all relevant timers, resets the state to [STATE_IDLE] and emits a JS error event
     *
     * - [RecognitionListener.onResults]:
     *   - Considered to be the end of the listening session (happening natively, so not stopped or canceled)
     *   and extracts the best final transcript from [SpeechRecognizer.RESULTS_RECOGNITION]
     *
     * - [RecognitionListener.onPartialResults]:
     *   - Similar to [RecognitionListener.onResults], but only runs when interim results are enabled
     *   ([interimResultsEnabled] == true)
     *   - Will try to extract the best segmented transcript
     *
     * - All other callbacks ([onReadyForSpeech], [onBeginningOfSpeech], [onRmsChanged],
     *   [onBufferReceived], [onEndOfSpeech], [onEvent]) are currently treated as no-ops
     *
     * This method should be called before starting a new dictation session to ensure
     * that a configured [SpeechRecognizer] instance is available.
     */

    private fun setupRecognizerIfNeeded() {
        if (speechRecognizer != null) return

        val ctx = reactContext.currentActivity ?: reactContext

        speechRecognizer = SpeechRecognizer.createSpeechRecognizer(reactContext).apply {
            setRecognitionListener(object : RecognitionListener {
                override fun onReadyForSpeech(params: Bundle?) = Unit
                override fun onBeginningOfSpeech() = Unit
                override fun onRmsChanged(rmsdB: Float) = Unit
                override fun onBufferReceived(buffer: ByteArray?) = Unit
                override fun onEndOfSpeech() = Unit

                override fun onError(error: Int) {
                    stopSilenceMonitor()
                    isListening.set(false)
                    emitAndroidError(error)
                    emitState(STATE_IDLE)
                }

                override fun onResults(results: Bundle?) {
                    stopSilenceMonitor()
                    isListening.set(false)

                    val text = results
                        ?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
                        ?.firstOrNull()
                        ?: ""

                    if (text.isNotBlank()) {
                        emitResult(text, true)
                    }
                    emitState(STATE_IDLE)
                }

                override fun onPartialResults(partialResults: Bundle?) {
                    if (!interimResultsEnabled) return

                    val text = partialResults
                        ?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
                        ?.firstOrNull()
                        ?: ""

                    if (text.isNotBlank()) {
                        lastSpeechTimeMs = System.currentTimeMillis()
                        emitResult(text, false)
                    }
                }

                override fun onEvent(eventType: Int, params: Bundle?) = Unit
            })
        }
    }

    /**
     * Fully tears down the current [SpeechRecognizer] instance and resets dictation state.
     *
     * This should be called when we are done with speech recognition for the current
     * lifecycle of the host (e.g. `onHostDestroy`) or when we want to guarantee that
     * no recognizer resources are left hanging around.
     */
    private fun cleanupRecognizer() {
        stopSilenceMonitor()
        mainHandler.post {
            try {
                speechRecognizer?.setRecognitionListener(null)
                speechRecognizer?.destroy()
            } catch (_: Throwable) {
                // ignore
            } finally {
                speechRecognizer = null
                isListening.set(false)
                emitState(STATE_IDLE)
            }
        }
    }

    /**
     * Returns `true` if the app has the RECORD_AUDIO permission.
     */
    private fun hasAudioPermission(): Boolean {
        val granted = ContextCompat.checkSelfPermission(
            reactContext,
            Manifest.permission.RECORD_AUDIO,
        )
        return granted == PackageManager.PERMISSION_GRANTED
    }

    /**
     * Starts (or restarts) a rudimentary silence monitor.
     *
     * Periodically checks the time since the last speech activity and if the
     * elapsed time exceeds [silenceTimeoutMs] while still listening, calls [stop].
     */
    private fun startSilenceMonitor() {
        stopSilenceMonitor()

        val runnable = object : Runnable {
            override fun run() {
                if (!isListening.get()) return

                val elapsed = System.currentTimeMillis() - lastSpeechTimeMs
                if (elapsed >= silenceTimeoutMs) {
                    stop()
                    return
                }

                mainHandler.postDelayed(this, 300L)
            }
        }

        silenceRunnable = runnable
        mainHandler.postDelayed(runnable, 300L)
    }

    /**
     * Stops the silence monitor by removing any pending callbacks.
     */
    private fun stopSilenceMonitor() {
        silenceRunnable?.let { mainHandler.removeCallbacks(it) }
        silenceRunnable = null
    }

    /**
     * Emits a dictation result event to JS.
     *
     * @param text    Transcribed text
     * @param isFinal Whether this result is final (`true`) or interim (`false`)
     */
    private fun emitResult(text: String, isFinal: Boolean) {
        val map = Arguments.createMap().apply {
            putString("text", text)
            putBoolean("isFinal", isFinal)
        }
        sendEvent(EVENT_RESULT, map)
    }

    /**
     * Emits a state change event to JS (e.g. IDLE, LISTENING etc).
     */
    private fun emitState(state: String) {
        val map = Arguments.createMap().apply {
            putString("state", state)
        }
        sendEvent(EVENT_STATE, map)
    }

    /**
     * Emits a generic error event to JS with a code and message.
     */
    private fun emitError(code: String, message: String) {
        val map = Arguments.createMap().apply {
            putString("code", code)
            putString("message", message)
        }
        sendEvent(EVENT_ERROR, map)
    }

    /**
     * Maps an Android SpeechRecognizer error code to a human-readable message
     * and emits it as a JS error event.
     */
    private fun emitAndroidError(errorCode: Int) {
        val msg = when (errorCode) {
            SpeechRecognizer.ERROR_AUDIO -> "Audio recording error"
            SpeechRecognizer.ERROR_CLIENT -> "Client side error"
            SpeechRecognizer.ERROR_INSUFFICIENT_PERMISSIONS -> "Insufficient permissions"
            SpeechRecognizer.ERROR_NETWORK -> "Network error"
            SpeechRecognizer.ERROR_NETWORK_TIMEOUT -> "Network timeout"
            SpeechRecognizer.ERROR_NO_MATCH -> "No recognition result matched"
            SpeechRecognizer.ERROR_RECOGNIZER_BUSY -> "RecognitionService busy"
            SpeechRecognizer.ERROR_SERVER -> "Server error"
            SpeechRecognizer.ERROR_SPEECH_TIMEOUT -> "No speech input"
            else -> "Unknown speech recognizer error ($errorCode)"
        }

        emitError(ERROR_ANDROID, msg)
    }

    /**
     * Sends a device event to JS via RCTDeviceEventEmitter.
     *
     * Safely no-ops if there is no active React instance or if emission throws.
     */
    private fun sendEvent(eventName: String, params: WritableMap?) {
        if (!reactContext.hasActiveReactInstance()) return
        try {
            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(eventName, params)
        } catch (e: Throwable) {
            throw e;
            // "Swallow" the error here if something throws. This
            // will typically happen if the React instance is
            // "falsely active" or if emission errors out. In both
            // scenarios, we don't want to crash native because JS
            // is missing.
        }
    }

    private fun ReadableMap.getBooleanSafe(key: String): Boolean? =
        if (hasKey(key) && getType(key) == ReadableType.Boolean) getBoolean(key) else null

    private fun ReadableMap.getStringSafe(key: String): String? =
        if (hasKey(key) && getType(key) == ReadableType.String) getString(key) else null

    private fun ReadableMap.getDoubleSafe(key: String): Double? =
        if (hasKey(key) && getType(key) == ReadableType.Number) getDouble(key) else null
}

File: packages/react-native-sdk/ios/AIDictationModule.swift
import Foundation
import Speech
import AVFoundation
import React

/**
 * A sink for dictation events emitted from the native module.
 *
 * Implemented on the Obj-C++ TurboModule side (NativeAIDictationSpecBase subclass),
 * which forwards these calls into React Native’s event system via codegen.
 *
 * All payloads are plain `NSDictionary` objects so they can be passed directly
 * to the generated `emitOn<name>` helpers on iOS.
 */
@objc public protocol AIDictationEventSink {
    func emitOnState(_ value: NSDictionary)
    func emitOnResult(_ value: NSDictionary)
    func emitOnError(_ value: NSDictionary)
}

@objc(AIDictationModule)
public class AIDictationModule: NSObject {

    // -------------------------------------------------------------------------
    // Constants
    // -------------------------------------------------------------------------

    /**
     * The logical module name exposed to JS.
     *
     * This must stay in sync with the TurboModule spec name so that
     * `NativeAIDictation` (JS) binds to this native implementation.
     */
    static let NAME = "NativeAIDictation"

    // High-level dictation state labels. These are emitted to JS via
    // the `state` field as part of a `type: "state"` event payload.
    private let STATE_IDLE      = "idle"
    private let STATE_STARTING  = "starting"
    private let STATE_LISTENING = "listening"
    private let STATE_STOPPING  = "stopping"

    // Error codes used in `type: "error"` events.
    private let ERROR_NO_PERMISSION = "NO_PERMISSION"
    private let ERROR_NOT_AVAILABLE = "NOT_AVAILABLE"
    private let ERROR_INTERNAL      = "INTERNAL_ERROR"

    /// Default maximum allowed silence (in milliseconds) before auto-stopping.
    private let DEFAULT_SILENCE_TIMEOUT_MS: Double = 2500.0

    // -------------------------------------------------------------------------
    // Internal state
    // -------------------------------------------------------------------------

    // AVFoundation/SpeechKit primitives driving the recognition session.
    private let audioEngine = AVAudioEngine()
    private var speechRecognizer: SFSpeechRecognizer?
    private var recognitionRequest: SFSpeechAudioBufferRecognitionRequest?
    private var recognitionTask: SFSpeechRecognitionTask?

    // Runtime configuration and state of the current dictation session.
    private var isListening: Bool = false
    private var interimResultsEnabled: Bool = true
    private var currentLocale: Locale = .current
    private var silenceTimeoutMs: Double = 2500.0
    private var lastSpeechTimeMs: Double = 0.0

    // Timer used for the simple silence monitor.
    private var silenceTimer: Timer?

    // -------------------------------------------------------------------------
    // Event emitter plumbing
    // -------------------------------------------------------------------------

    /**
     * Event sink provided by the Obj-C TurboModule wrapper.
     *
     * The Obj-C class sets this to itself and forwards
     * calls into the codegen-generated `emitOn<name>`
     * helpers, which in turn send events to JS.
     */
    @objc public weak var eventSink: AIDictationEventSink?

    /**
     * Convenience entry point wired from Obj-C if needed.
     *
     * This signature exists so Objective-C can call
     * `[_impl startWithOptions:options eventEmitter:self]` without having
     * to construct Swift-specific types. Internally it just forwards to
     * the primary `start(_:)` implementation.
     */
    @objc(startWithOptions:eventEmitter:)
    public func startWithOptions(_ options: NSDictionary?, eventEmitter: NSObject) {
        start(options)
    }

    // -------------------------------------------------------------------------
    // NativeAIDictationSpec API
    // -------------------------------------------------------------------------

    /**
     * Starts a new dictation session with the given options.
     *
     * - No-ops if a session is already active
     * - Validates microphone + speech recognition permissions
     * - On success, configures and starts `AVAudioEngine` + `SFSpeechRecognizer`
     *   and arms a basic silence monitor
     *
     * Expected `options` keys (all optional):
     *
     * - `interimResults` (`Bool`, default `true`):
     *      Whether to emit partial / interim results while the user is speaking.
     *
     * - `language` (`String`, BCP-47 tag):
     *      Locale identifier (e.g. `"en-US"`). Falls back to `Locale.current`
     *      if omitted or blank.
     *
     * - `silenceTimeoutMs` (`Double`):
     *      Maximum allowed silence, in milliseconds, before auto-stopping.
     *      Falls back to `DEFAULT_SILENCE_TIMEOUT_MS` if missing or invalid.
     */
    @objc
    public func start(_ options: NSDictionary?) {
        if isListening {
            return
        }

        emitState(STATE_STARTING)

        // Runtime configuration derived from options.
        interimResultsEnabled = (options?["interimResults"] as? Bool) ?? true

        if let languageTag = options?["language"] as? String,
           !languageTag.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            currentLocale = Locale(identifier: languageTag)
        } else {
            currentLocale = .current
        }

        if let silenceOpt = options?["silenceTimeoutMs"] as? Double, silenceOpt > 0 {
            silenceTimeoutMs = silenceOpt
        } else {
            silenceTimeoutMs = DEFAULT_SILENCE_TIMEOUT_MS
        }

        // Permissions
        if !hasAudioPermission() {
            // If permissions are missing, request them and resolve via callback.
            requestPermissions { granted in
                if !granted {
                    self.emitError(
                        self.ERROR_NO_PERMISSION,
                        message: "Microphone or speech recognition permission not granted."
                    )
                    self.emitState(self.STATE_IDLE)
                }
            }
            return
        }

        // Availability
        guard let recognizer = SFSpeechRecognizer(locale: currentLocale), recognizer.isAvailable else {
            emitError(
                ERROR_NOT_AVAILABLE,
                message: "Speech recognition is not available for this locale or on this device."
            )
            emitState(STATE_IDLE)
            return
        }

        speechRecognizer = recognizer

        // Configure and start the audio engine + recognition task on the main thread.
        DispatchQueue.main.async {
            do {
                try self.setupAudioSession()
                try self.startRecognition()
            } catch {
                self.isListening = false
                self.emitState(self.STATE_IDLE)
                self.emitError(
                    self.ERROR_INTERNAL,
                    message: "Failed to start recognition: \(error.localizedDescription)"
                )
            }
        }
    }

    /**
     * Requests a graceful stop of the current dictation session.
     *
     * - No-ops if we are not currently listening
     * - Calls `endAudio()` on the recognition request, allowing
     *   `SFSpeechRecognizer` to deliver final results before terminating
     *
     * Final state transitions are handled via the recognition task callback
     * (`finishRecognitionSession()` or the error branch).
     */
    @objc
    public func stop() {
        if !isListening {
            return
        }

        emitState(STATE_STOPPING)
        stopSilenceMonitor()

        DispatchQueue.main.async {
            self.recognitionRequest?.endAudio()
        }
    }

    /**
     * Immediately cancels the current dictation session.
     *
     * - No-ops if we are not currently listening
     * - Stops the audio engine and cancels the recognition task
     * - Resets all internal state and emits `STATE_IDLE`
     *
     * Use this when the session should be aborted (e.g. user cancelled) rather
     * than gracefully completed.
     */
    @objc
    public func cancel() {
        if !isListening {
            return
        }

        stopSilenceMonitor()

        DispatchQueue.main.async {
            self.stopAudioEngine()
            self.recognitionTask?.cancel()
            self.recognitionTask = nil
            self.recognitionRequest = nil
            self.isListening = false
            self.emitState(self.STATE_IDLE)
        }
    }

    /**
     * Synchronously reports whether a dictation session is currently active.
     *
     * This is exposed to JS as a sync method so callers can quickly check
     * recording state without waiting on an event.
     */
    @objc
    public func isRecording() -> Bool {
        return isListening
    }

    // -------------------------------------------------------------------------
    // Internal setup / teardown
    // -------------------------------------------------------------------------

    /**
     * Configures the shared `AVAudioSession` for recording.
     *
     * - Uses `.record` category and `.measurement` mode for best recognition
     *   quality.
     * - Ducks other audio and notifies other sessions on deactivation.
     *
     * Throws if the session cannot be configured or activated.
     */
    private func setupAudioSession() throws {
        let session = AVAudioSession.sharedInstance()
        try session.setCategory(.record, mode: .measurement, options: [.duckOthers])
        try session.setActive(true, options: .notifyOthersOnDeactivation)
    }

    /**
     * Starts a fresh `SFSpeechRecognitionTask` and connects it to `AVAudioEngine`.
     *
     * Responsibilities:
     * - Cleans up any previous recognition task
     * - Creates a new `SFSpeechAudioBufferRecognitionRequest`
     * - Installs an input node tap to feed audio buffers into the request
     *
     * The recognition task callback:
     * - Emits interim and final transcripts via `emitResult(...)`
     * - Advances the silence timer on non-empty results
     * - On final results, calls `finishRecognitionSession()`
     * - On error, emits an error event, resets state, and tears down the
     *   engine / task
     */
    private func startRecognition() throws {
        // Clean up any previous task
        recognitionTask?.cancel()
        recognitionTask = nil

        let request = SFSpeechAudioBufferRecognitionRequest()
        request.shouldReportPartialResults = interimResultsEnabled
        recognitionRequest = request

        let inputNode = audioEngine.inputNode
        let recordingFormat = inputNode.outputFormat(forBus: 0)

        inputNode.removeTap(onBus: 0)
        inputNode.installTap(onBus: 0, bufferSize: 1024, format: recordingFormat) {
            [weak self] (buffer, _) in
            self?.recognitionRequest?.append(buffer)
        }

        audioEngine.prepare()
        try audioEngine.start()

        lastSpeechTimeMs = currentTimeMs()
        isListening = true
        emitState(STATE_LISTENING)

        startSilenceMonitor()

        recognitionTask = speechRecognizer?.recognitionTask(with: request) { [weak self] result, error in
            guard let self = self else { return }

            if let result = result {
                let text = result.bestTranscription.formattedString
                if !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    self.lastSpeechTimeMs = self.currentTimeMs()
                    self.emitResult(text: text, isFinal: result.isFinal)
                }

                if result.isFinal {
                    self.finishRecognitionSession()
                }
            }

            if let error = error {
                self.stopSilenceMonitor()
                self.isListening = false
                self.emitError(
                    self.ERROR_INTERNAL,
                    message: "Speech recognizer error: \(error.localizedDescription)"
                )
                self.emitState(self.STATE_IDLE)
                self.stopAudioEngine()
                self.recognitionTask = nil
                self.recognitionRequest = nil
            }
        }
    }

    /**
     * Handles the normal end-of-session flow after a final recognition result.
     *
     * - Stops the silence monitor
     * - Marks `isListening = false`
     * - Emits `STATE_IDLE`
     * - Stops the audio engine and clears recognition objects
     */
    private func finishRecognitionSession() {
        stopSilenceMonitor()
        isListening = false
        emitState(STATE_IDLE)
        stopAudioEngine()
        recognitionTask = nil
        recognitionRequest = nil
    }

    /**
     * Fully tears down the recognizer and audio engine and resets state.
     *
     * Intended for hard cleanup paths (e.g. deinit, React instance teardown).
     * Ensures we:
     * - stop the silence monitor
     * - stop and detach taps from the audio engine
     * - cancel and nil out the recognition task / request
     * - release the `SFSpeechRecognizer` instance
     * - emit `STATE_IDLE`
     */
    private func cleanupRecognizer() {
        stopSilenceMonitor()
        DispatchQueue.main.async {
            self.stopAudioEngine()
            self.recognitionTask?.cancel()
            self.recognitionTask = nil
            self.recognitionRequest = nil
            self.speechRecognizer = nil
            self.isListening = false
            self.emitState(self.STATE_IDLE)
        }
    }

    /**
     * Stops the audio engine and removes any installed tap on the input node.
     */
    private func stopAudioEngine() {
        if audioEngine.isRunning {
            audioEngine.stop()
            audioEngine.inputNode.removeTap(onBus: 0)
        }
    }

    // -------------------------------------------------------------------------
    // Permissions / silence monitor
    // -------------------------------------------------------------------------

    /**
     * Returns `true` if both microphone and speech recognition permissions
     * are already granted.
     *
     * This is a simple synchronous check used before starting recognition.
     */
    private func hasAudioPermission() -> Bool {
        let micAllowed = AVAudioSession.sharedInstance().recordPermission == .granted
        let speechAllowed = SFSpeechRecognizer.authorizationStatus() == .authorized
        return micAllowed && speechAllowed
    }

    /**
     * Requests microphone + speech recognition permissions.
     *
     * The completion handler is invoked on the main thread with `true`
     * only if both:
     * - `AVAudioSession` record permission is granted
     * - `SFSpeechRecognizer` authorization status is `.authorized`
     */
    private func requestPermissions(completion: @escaping (Bool) -> Void) {
        let audioSession = AVAudioSession.sharedInstance()

        audioSession.requestRecordPermission { micGranted in
            SFSpeechRecognizer.requestAuthorization { status in
                let speechGranted = (status == .authorized)
                DispatchQueue.main.async {
                    completion(micGranted && speechGranted)
                }
            }
        }
    }

    /**
     * Starts (or restarts) a rudimentary silence monitor.
     *
     * Periodically checks the time since the last speech activity and if the
     * elapsed time exceeds `silenceTimeoutMs` while still listening, calls `stop()`.
     *
     * The timer runs on the main run loop in `.common` mode so that it continues
     * to fire during typical UI interactions.
     */
    private func startSilenceMonitor() {
        stopSilenceMonitor()

        guard silenceTimeoutMs > 0 else { return }

        silenceTimer = Timer.scheduledTimer(withTimeInterval: 0.3, repeats: true) { [weak self] _ in
            guard let self = self, self.isListening else { return }

            let elapsed = self.currentTimeMs() - self.lastSpeechTimeMs
            if elapsed >= self.silenceTimeoutMs {
                self.stop()
            }
        }
        if let timer = silenceTimer {
            RunLoop.main.add(timer, forMode: .common)
        }
    }

    /**
     * Stops the silence monitor by invalidating any pending timer.
     */
    private func stopSilenceMonitor() {
        silenceTimer?.invalidate()
        silenceTimer = nil
    }

    /**
     * Returns the current wall-clock time in milliseconds.
     *
     * Used to track the last time non-empty speech was observed so that the
     * silence monitor can auto-stop idle sessions.
     */
    private func currentTimeMs() -> Double {
        return Date().timeIntervalSince1970 * 1000.0
    }

    // -------------------------------------------------------------------------
    // Event helpers
    // -------------------------------------------------------------------------

    /**
     * Emits a dictation result event to JS.
     *
     * Payload fields:
     * - `text` (`String`): recognized transcript
     * - `isFinal` (`Bool`): `true` for final results, `false` for interim
     */
    private func emitResult(text: String, isFinal: Bool) {
        let payload: NSDictionary = [
            "text": text,
            "isFinal": isFinal
        ]
        eventSink?.emitOnResult(payload)
    }

    /**
     * Emits a state change event to JS.
     *
     * Payload fields:
     * - `state` (`String`): one of `"idle"`, `"starting"`, `"listening"`, `"stopping"`
     */
    private func emitState(_ state: String) {
        let payload: NSDictionary = [
            "state": state
        ]
        eventSink?.emitOnState(payload)
    }

    /**
     * Emits a generic error event to JS with a code and message.
     *
     * Payload fields:
     * - `code` (`String`): internal error code (e.g. `NO_PERMISSION`, `NOT_AVAILABLE`)
     * - `message` (`String`): human-readable description for logging / debugging
     */
    private func emitError(_ code: String, message: String) {
        let payload: NSDictionary = [
            "code": code,
            "message": message
        ]
        eventSink?.emitOnError(payload)
    }
}
